GVFS/GVFS.Common/Git/LibGit2RepoInvoker.cs

```diff
@@ -0,0 +1,118 @@
+﻿using GVFS.Common.Tracing;
+using System;
+using System.Threading;
+
+namespace GVFS.Common.Git
+{
+    public class LibGit2RepoInvoker : IDisposable
+    {
+        private static readonly TimeSpan DefaultRepositoryDisposalPeriod = TimeSpan.FromMinutes(15);
+
+        private readonly Func<LibGit2Repo> createRepo;
+        private readonly ITracer tracer;
+        private readonly object sharedRepoLock = new object();
+        private readonly TimeSpan sharedRepositoryDisposalPeriod;
+        private volatile bool stopped;
+        private volatile int activeCallers;
+        private LibGit2Repo sharedRepo;
+        private Timer sharedRepoDisposalTimer;
+
+        public LibGit2RepoInvoker(ITracer tracer, Func<LibGit2Repo> createRepo, TimeSpan? disposalPeriod = null)
+        {
+            this.tracer = tracer;
+            this.createRepo = createRepo;
+
+            if (!disposalPeriod.HasValue || disposalPeriod.Value <= TimeSpan.Zero)
+            {
+                this.sharedRepositoryDisposalPeriod = DefaultRepositoryDisposalPeriod;
+            }
+            else
+            {
+                this.sharedRepositoryDisposalPeriod = disposalPeriod.Value;
+            }
+
+            this.sharedRepoDisposalTimer = new Timer(
+                (state) => this.DisposeSharedRepo(),
+                state: null,
+                dueTime: this.sharedRepositoryDisposalPeriod,
+                period: this.sharedRepositoryDisposalPeriod);
+        }
+
+        public bool IsActive => this.sharedRepo != null;
+
+        public void Dispose()
+        {
+            this.stopped = true;
+
+            this.sharedRepoDisposalTimer?.Dispose();
+            this.sharedRepoDisposalTimer = null;
+
+            lock (this.sharedRepoLock)
+            {
+                this.sharedRepo?.Dispose();

```

Comment:
Do we need to check the `this.activeCallers` like in `DisposeSharedRepo`?

Reply:
This comment is interesting, and I haven't done anything about it in the code yet.

This `Dispose()` method is only called when we are shutting down the process. So someone would need to be hydrating at the same time they are running `gvfs unmount`. I'm pretty sure that already has unclear behavior, but we don't want to make it worse.

Perhaps the safest thing to do is to skip disposing the shared repo if there are active callers? It will get cleaned up as the process shuts down, so it's a safer response to that situation than the current behavior.

Thoughts?

Reply:
Yeah the one thing I wanted to avoid was a crash because Dispose was called but somehow there was an active request to invoke that runs and we get either a disposed or null object exception because we would probably get support issues from people that ran `gvfs unmount` and had a popup for the crash or would see it in the event viewer.